File: src/WorkflowShell/adb/SendText.rb
require_relative '../models/Command'

class SendText < Command
  @command_string
  @command_description
  @command_usage

  def initialize
    @command_string = 'ast'
    @command_description = 'Sends text with ADB.'
    @command_usage = @command_string
  end

  def run_command(arguments)
    # Ask the Command class to parse the basic options like help and verbosity
    parser_components = parse_options(arguments, false)
    if parser_components.nil?
      puts "You MUST specify the text to send!"
      puts ""
      print_help(parser_components.basic_options_printer, nil)
    end

    text_to_send = parser_components.leftover_arguments.pop
    text_to_send = text_to_send.gsub('(','\(')
    text_to_send = text_to_send.gsub(')','\)')
    text_to_send = text_to_send.gsub('<','\<')
    text_to_send = text_to_send.gsub('>','\>')
    text_to_send = text_to_send.gsub('|','\|')
    text_to_send = text_to_send.gsub(';','\;')
    text_to_send = text_to_send.gsub('&','\&')
    text_to_send = text_to_send.gsub('*','\*')
    text_to_send = text_to_send.gsub('~','\~')
    text_to_send = text_to_send.gsub('\"','\\\"')
    text_to_send = text_to_send.gsub('\'','\\\'')
    text_to_send = text_to_send.gsub(' ','%s')

    command = "adb shell input text \"#{text_to_send}\""
    puts "Command to send: " + command
    run_shell_command(command, parser_components.basic_options.verbose)
  end
end

File: src/WorkflowShell/adb/SendText.ts
import Command from "../models/Command";
import Logger from "../util/Logger";
const logger = new Logger("SendText");

export default class SendText extends Command {

  static getString(): string {
    return "ast";
  }

  static getDescription(): string {
    return "Sends text with ADB.";
  }

  getUsage(): string {
    return `${SendText.getString()} <Message>`;
  }

  async run(args: any) {
    const ok: boolean = await super.run(args);
    if (!ok) {
      return false;
    }

    let textToSend: string = args._.shift();
    if (!textToSend) {
      logger.log("You MUST specify the text to send!");
      this.printHelp();
      return false;
    }

    textToSend = textToSend.replace(/\(/g, "\(");
    textToSend = textToSend.replace(/\)/g, "\)");
    textToSend = textToSend.replace(/</g, "\<");
    textToSend = textToSend.replace(/>/g, "\>");
    textToSend = textToSend.replace(/\|/g, "\|");
    textToSend = textToSend.replace(/;/g, "\;");
    textToSend = textToSend.replace(/&/g, "\&");
    textToSend = textToSend.replace(/\*/g, "\*");
    textToSend = textToSend.replace(/~/g, "\~");
    textToSend = textToSend.replace(/"/g, "\\\"");
    textToSend = textToSend.replace(/'/g, "\\'");
    textToSend = textToSend.replace(/ /g, "%s");

    const baseCommand: string = `adb shell input text "${textToSend}"`;
    super.execute(baseCommand);
    return true;
  }
}
